docs/CompilerPerformance.md

```diff
@@ -0,0 +1,1142 @@
+  - Expression type inference solves constraints inefficiently, and can
+    sometimes behave super-linearly or even exponentially.
+  - Periodically the analysis phase of a SIL optimization fails to cache
+    overlapping subproblems, causing a super-linear slowdown.
+  - Some SIL-to-IR lowerings (eg. large value types, ARC operations) can
+    generate too much LLVM IR, increasing the time spent in LLVM.
+
+(Subsystem experts: please add further areas of concern here.)
+
+## How to diagnose compilation performance problems
+
+Compiler performance analysis breaks down into two broad categories of work,
+depending on what you're trying to do:
+
+  - Isolating a regression
+  - Finding areas that need general improvement
+
+In all cases, it's important to be familiar with several tools and compiler
+options we have at our disposal. If you know about all these tools, you can skip
+the following section.
+
+### Tools and options
+
+You'll use several tools along the way. These come in 5 main categories:
+
+  - Profilers
+  - Diagnostic options built-in to the compiler (timers, counters)
+  - Post-processing tools to further analyze diagnostic output
+  - Tools to generally analyze the output artifacts of the compiler
+  - Tools to minimize the regression range or testcases
+
+#### Profilers
+
+The basic tool of performance analysis is a profiler, and you _will need_ to
+learn to use at least one profiler for the purposes of this work. The main two
+profilers we use are `Instruments.app` on macOS, and `perf(1)` on Linux. Both
+are freely available and extremely powerful; this document will barely scratch
+the surface of what they can do.
+
+##### Instruments.app
+
+Instruments is a tool on macOS that ships as part of Xcode. It contains
+graphical and batch interfaces to a very wide variety of profiling services;
+see
+[here](https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/InstrumentsUserGuide/index.html) for
+more documentation.
+
+The main way we will use `Instruments.app` is in "Counter" mode, to record and
+analyze a single run of swiftc. We will also use it in simple push-button
+interactive mode, as a normal application. While it's possible to run
+Instruments in batch mode on the command-line, the batch interface is less
+reliable than running it as an interactive application, and frequently causes
+lockups or fails to collect data.
+
+Before starting, you should also be sure you are going to profile a version of
+Swift _without_ DWARF debuginfo; while in theory debuginfo will give a
+higher-resolution, more-detailed profile, in practice Instruments will often
+stall out and become unresponsive trying to process the additional detail.
+
+Similarly, be sure that as many applications as possible (especially those with
+debuginfo themselves!) are closed, so that Instruments has has little additional
+material to symbolicate as possible. It collects a _whole system profile_ at
```

Comment:
It may be good to mention that you can use instruments on an individual process, but generally profiling the system works well enough.

Reply:
Yeah there are a ton of ways instruments can theoretically be used in a more-precise fashion (command line, single-launched-process, etc.) none of which I can get to work reliably. It sometimes works, but usually misses the process entirely, reports "process exited" or no data at all.